packages/admin/cms-admin/src/dependencies/createDocumentDependencyMethods.ts

```diff
@@ -0,0 +1,98 @@
+
+import { GQLPageTreeNode, Maybe } from "../graphql.generated";
+import { DependencyInterface } from "./types";
+
+interface Query<RootBlocks extends Record<string, BlockInterface>> {
+    node: Maybe<
+        { id: string; pageTreeNode: Maybe<Pick<GQLPageTreeNode, "id" | "category">> } & { [Key in keyof RootBlocks]: BlockInputApi<RootBlocks[Key]> }
+    >;
+}
+
+interface QueryVariables {
+    id: string;
+}
+
+export function createDocumentDependencyMethods<RootBlocks extends Record<string, BlockInterface>>({
+    rootQueryName,
+    rootBlocks,
+    basePath,
+}: {
+    rootQueryName: string;
+    rootBlocks: { [Key in keyof RootBlocks]: RootBlocks[Key] | { block: RootBlocks[Key]; path?: string } };
+    basePath:
+        | string
+        | ((
+              node: { id: string; pageTreeNode: Pick<GQLPageTreeNode, "id" | "category"> } & {
+                  [Key in keyof RootBlocks]: BlockInputApi<RootBlocks[Key]>;
+              },
+          ) => string);
+}): Pick<DependencyInterface, "resolveRoute"> {
+    return {
+        resolveRoute: async ({ rootColumnName, jsonPath, apolloClient, id }) => {
+            const { data, error } = await apolloClient.query<Query<RootBlocks>, QueryVariables>({
+                query: gql`
+                    query ${rootQueryName}Dependency($id: ID!) {
+                        node: ${rootQueryName}(id: $id) {
+                            id
+                            ${Object.keys(rootBlocks).join("\n")}
+                            pageTreeNode {
+                                id
+                                category
+                            }
+                        }    
+                    }    
+                `,
+                variables: {
+                    id,
+                },
+            });
+
+            if (error || data.node === null) {
+                throw new Error(`Error for document ${id}: ${error?.message ?? "Document is undefined"}`);
+            }
+
+            if (!hasPageTreeNode(data.node)) {
+                throw new Error(`Error for document ${id}: No page tree node found`);
+            }
+
+            let url: string;
+
+            if (typeof basePath === "string") {
+                url = basePath;
+            } else {
+                url = basePath(data.node);
+            }
+
+            if (jsonPath && rootColumnName) {
+                let block: BlockInterface;
+                let path: string | undefined;
+
+                const rootBlock = rootBlocks[rootColumnName];
+
+                if (isBlockInterface(rootBlock)) {
+                    block = rootBlock;
+                } else {
+                    block = rootBlock.block;
+                    path = rootBlock.path;
+                }
+
+                url += `${path ?? ""}/`;
+                url += block.resolveDependencyRoute(block.input2State(data.node[rootColumnName]), jsonPath.substring("root.".length));
```

Comment:
We should decide whether we're talking about a path or a route, currently we have both.

Reply:
IMO it's a path, and we should rename the methods `resolvePath` and `resolveDependencyPath`

Reply:
I'm fine with both. Feel free to rename it to path

Reply:
I will do that in a follow-up PR